packages/gatsby-plugin-fastify/src/plugins/static.ts

```diff
@@ -16,15 +15,7 @@ export const handleStatic: FastifyPluginAsync<Partial<FastifyStaticOptions>> = f
       // These settings were switched to false with fastify v4, not entirely sure what changed, but tests are still passing.
       redirect: false,
       wildcard: false,
-      setHeaders: (reply, path) => {
-        if (
-          isMatch(path, ["**/public/*.@(js|css)", "**/public/static/**"]) &&
-          isMatch(path, "!**/sw.js")
-        ) {
-          reply.setHeader(...IMMUTABLE_CACHING_HEADER);
```

Comment:
The `IMMUTABLE_CACHING_HEADER` is added in two spots.

In `constants.ts`, it is added to `/static/**` this is equivalent to the above `**/public/static/**` since the pattern matching will now be done on the `request.url` vs. the static file path that `fastify-static` passes in. 

```
export const CACHING_HEADERS = {
  "/static/**": IMMUTABLE_CACHING_HEADER,
```

As well, in `header-builder.ts` it is added to all chunked files consistent with gatsby hosting. 

I'm not sure where this rule came from `"**/public/*.@(js|css)"`. I believe only chunked files should be `immutable`, not all `.js`, `.css` file within the root. I noticed this on my own site where generated files in the root for lazy loaded components do not get a chunk entry and gatsby hosting does not set any `cache-control` headers at all for them. 

In order to reproduce and test this behaviour, I added a lazy loaded component as well as support for "unsetting" headers in 8fe0bc0fe4a595d59f481e25fdb47e343c5cd619 which results in both gatsby hosting and the fastify server using the same "unset" `cache-control` header for these `.js` files as opposed to an immutable header:
![image](https://user-images.githubusercontent.com/4126987/225120112-636d5b72-9e83-40f1-ba1b-024288d87fbc.png)

For `.css` files in the root, I'm not sure how to reproduce actually including a CSS file from the root as importing any CSS gets added to the CSS chunk and served as a chunked file which gets the appropriate immutable header due to being chunked.

Reply:
Ooooookay...I see what's happening. Thanks for the explanation.  I think this reinforces my desire for this to all happen at build time. Right now it seems that the application of default and custom headers is split between build-time (`gatsby/header-build.ts`), and runtime (`plugins/headers.ts). From a code perspective, this seems confusing, not to mention the potential performance issues. 

So, the solution is to enumerate all possible routes at build time (already done)...and use those to calculate headers for that route. This can be written to the config and used at runtime to actually apply those headers to each route in `plugins/headers.ts`.

This generally applies to all route paterns `/example/[post_slug].ts`. The major exception being if the header applies to `/example/test/`. We don't know that that exists but we will assume it does and save the header in the lookup table. That may take some special logic. lmk if you have any questions.

As for CSS files, nice catch on the CSS and JS files. IIRC that was an intentional simplification, though technically a bug. If I'm not wrong, the only JS and CSS files NOT hashed would be those included in the `static` folder. Are lazy loaded components also an exception? Atleast for CSS, the way to test would be adding a test fil in `src/static`, this will get coppied to `public` without a hash I believe.

Reply:
Resolved in ab6ff63f7c327471f681e9c067ea236a58f52163